File: java/com/example/spottio/ProfileFragment.java
package com.example.spottio;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

public class ProfileFragment extends Fragment {

    private String targetUser;
    private String currentUser;

    private ProfileViewModel viewModel;
    private ProfileUIManager uiManager;
    private ProfileCoordinator coordinator;

    private ProfileViewBinder viewBinder;
    private ProfileImageHandler imageHandler;

    public ProfileFragment() {}

    public static ProfileFragment newInstance(String targetUser) {
        // JAVA 17: Uso di 'var' per istanziare oggetti senza ripetere il tipo
        var fragment = new ProfileFragment();
        var args = new Bundle();
        args.putString("ARG_TARGET_USER", targetUser);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) targetUser = getArguments().getString("ARG_TARGET_USER");

        currentUser = AuthManager.getCurrentUserUid(requireContext());
        if (currentUser.isEmpty()) {
            currentUser = "Guest";
        }

        // JAVA 17: 'var' rende più snella la lettura di lunghe chiamate di sistema
        var prefs = requireActivity().getSharedPreferences("SpottioPrefs", Context.MODE_PRIVATE);

        viewModel = new ViewModelProvider(this).get(ProfileViewModel.class);
        uiManager = new ProfileUIManager(requireContext());
        coordinator = new ProfileCoordinator(this, currentUser, targetUser);

        if (coordinator.isMyProfile()) {
            targetUser = currentUser;
        }

        imageHandler = new ProfileImageHandler();
        viewModel.checkAdminStatus(currentUser, prefs);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        // JAVA 17: 'var' per l'inflazione delle view
        var view = inflater.inflate(R.layout.fragment_profile, container, false);

        viewBinder = new ProfileViewBinder(view);
        viewBinder.rvMyPosts.setLayoutManager(new LinearLayoutManager(getContext()));

        imageHandler.bindImageView(viewBinder.ivPfp);
        viewBinder.setupLanguageSelector(requireContext(), coordinator);
        viewBinder.setupButtons(uiManager, coordinator, viewModel, currentUser, targetUser);
        viewBinder.setupClickListeners(coordinator, viewModel);

        observeViewModel();

        viewModel.loadProfileData(targetUser, currentUser, coordinator.isMyProfile());
        viewModel.loadUserPosts(targetUser);

        return view;
    }

    private void observeViewModel() {
        viewModel.getUsername().observe(getViewLifecycleOwner(), name -> viewBinder.setTargetUser(name));
        viewModel.getBio().observe(getViewLifecycleOwner(), bio -> viewBinder.tvBio.setText(bio));
        viewModel.getFollowersCount().observe(getViewLifecycleOwner(), count -> viewBinder.tvFollowersCount.setText(String.valueOf(count)));
        viewModel.getFollowingCount().observe(getViewLifecycleOwner(), count -> viewBinder.tvFollowingCount.setText(String.valueOf(count)));

        viewModel.getIsFollowing().observe(getViewLifecycleOwner(), isFollowing ->
                uiManager.updateFollowButton(viewBinder.btnFollow, isFollowing)
        );

        viewModel.getProfileImageUrl().observe(getViewLifecycleOwner(), url -> {
            uiManager.setProfileImage(viewBinder.ivPfp, url);
            imageHandler.setupProfileImageClick(uiManager, coordinator.isMyProfile(), url, viewModel, currentUser);
        });

        viewModel.getIsVerified().observe(getViewLifecycleOwner(), isVerified ->
                UserVerificationManager.setupVerificationBadge(viewBinder.ivVerifiedBadge, isVerified)
        );

        viewModel.getErrorMessage().observe(getViewLifecycleOwner(), msg -> {
            if (msg != null && !msg.isEmpty()) {
                Toast.makeText(getContext(), msg, Toast.LENGTH_SHORT).show();
            }
        });

        viewModel.getIsAdmin().observe(getViewLifecycleOwner(), isAdmin ->
                uiManager.setupAdminFeatures(isAdmin, viewBinder.btnViewReports, () -> coordinator.navigateToAdminReports())
        );

        viewModel.getUserPosts().observe(getViewLifecycleOwner(), posts -> {
            // JAVA 17: Semplifichiamo il recupero dello stato Admin
            var isAdminObj = viewModel.getIsAdmin().getValue();
            boolean userIsAdmin = isAdminObj != null ? isAdminObj : false;

            // JAVA 17: Creazione di interfacce anonime tramite var
            var profilePostListener = new PostAdapter.PostInteractionListener() {
                @Override
                public void onUserClick(String userId) {
                    if (!userId.equals(targetUser)) {
                        getParentFragmentManager().beginTransaction()
                                .replace(R.id.fragment_container, ProfileFragment.newInstance(userId))
                                .addToBackStack(null).commit();
                    }
                }

                @Override
                public void onEditClick(Post post, int position) {
                    PostDialogHelper.showEditPostDialog(requireContext(), post, (PostAdapter) viewBinder.rvMyPosts.getAdapter(), position);
                }

                @Override
                public void onCommentClick(Post post, int position) {
                    PostDialogHelper.showCommentsSheet(requireContext(), post, currentUser, (PostAdapter) viewBinder.rvMyPosts.getAdapter(), position);
                }

                @Override
                public void onReportClick(Post post) {
                    PostDialogHelper.showReportDialog(requireContext(), post, currentUser);
                }

                @Override
                public void onShareClick(Post post, String shareContent) {
                    var sendIntent = new android.content.Intent(android.content.Intent.ACTION_SEND);
                    sendIntent.putExtra(android.content.Intent.EXTRA_TEXT, shareContent);
                    sendIntent.setType("text/plain");
                    startActivity(android.content.Intent.createChooser(sendIntent, "Condividi post tramite:"));
                }

                @Override
                public void onImageZoomClick(View imageView, String mediaUri) {
                    var builder = new com.ablanco.zoomy.Zoomy.Builder(requireActivity())
                            .target(imageView).enableImmersiveMode(false).animateZooming(true);
                    builder.register();
                }
            };

            viewBinder.rvMyPosts.setAdapter(new PostAdapter(posts, currentUser, userIsAdmin, profilePostListener));
        });
    }
}

File: java/com/example/spottio/ProfileFragment.kt
package com.example.spottio

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.ablanco.zoomy.Zoomy

class ProfileFragment : Fragment() {

    private var targetUser: String? = null
    private lateinit var currentUser: String

    private lateinit var viewModel: ProfileViewModel
    private lateinit var uiManager: ProfileUIManager
    private lateinit var coordinator: ProfileCoordinator
    private lateinit var viewBinder: ProfileViewBinder
    private lateinit var imageHandler: ProfileImageHandler

    companion object {
        @JvmStatic
        fun newInstance(targetUser: String): ProfileFragment {
            return ProfileFragment().apply {
                arguments = Bundle().apply {
                    putString("ARG_TARGET_USER", targetUser)
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        targetUser = arguments?.getString("ARG_TARGET_USER")
        currentUser = AuthManager.getCurrentUserUid(requireContext())

        if (targetUser.isNullOrEmpty()) {
            targetUser = currentUser
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_profile, container, false)

        viewModel = ViewModelProvider(this)[ProfileViewModel::class.java]
        uiManager = ProfileUIManager(requireContext())
        coordinator = ProfileCoordinator(this, currentUser, targetUser)
        viewBinder = ProfileViewBinder(view)
        imageHandler = ProfileImageHandler()

        imageHandler.bindImageView(viewBinder.ivPfp)
        viewBinder.rvMyPosts.layoutManager = LinearLayoutManager(requireContext())

        val isMyProfile = coordinator.isMyProfile()
        uiManager.configureProfileLayout(isMyProfile, view)

        viewBinder.configureForUser(
            isMyProfile,
            coordinator,
            viewModel,
            currentUser,
            targetUser!!,
            uiManager
        )

        viewBinder.setupClickListeners(coordinator, viewModel)

        setupLanguageSelector(view)
        setupAdminReportsButton()

        observeViewModel()

        viewModel.loadUserProfile(targetUser!!, currentUser)

        return view
    }

    private fun setupLanguageSelector(rootView: View) {
        val languageManager = LanguageSelectorManager(requireContext(), rootView)
        languageManager.init(object : LanguageSelectorManager.LanguageSelectionListener {
            override fun onLanguageSelected(code: String, flag: String, name: String) {
                coordinator.changeLanguage(code, flag, name)
            }
        })
    }

    private fun setupAdminReportsButton() {
        if (AuthManager.isCurrentUserAdmin(requireContext()) && coordinator.isMyProfile()) {
            viewBinder.btnViewReports?.visibility = View.VISIBLE
            viewBinder.btnViewReports?.setOnClickListener {
                coordinator.navigateToAdminReports()
            }
        } else {
            viewBinder.btnViewReports?.visibility = View.GONE
        }
    }

    private fun observeViewModel() {
        viewModel.username.observe(viewLifecycleOwner) { name ->
            view?.findViewById<TextView>(R.id.tvProfileName)?.text = name
        }

        viewModel.bio.observe(viewLifecycleOwner) { bioText ->
            viewBinder.tvBio.text = bioText
        }

        viewModel.followersCount.observe(viewLifecycleOwner) { count ->
            viewBinder.tvFollowersCount.text = count.toString()
        }

        viewModel.followingCount.observe(viewLifecycleOwner) { count ->
            viewBinder.tvFollowingCount.text = count.toString()
        }

        viewModel.isFollowing.observe(viewLifecycleOwner) { isFollowing ->
            uiManager.updateFollowButton(isFollowing, viewBinder.btnFollow)
        }

        viewModel.profileImageUrl.observe(viewLifecycleOwner) { url ->
            uiManager.loadProfileImage(url, viewBinder.ivPfp)
            imageHandler.setupProfileImageClick(
                uiManager,
                coordinator.isMyProfile(),
                url,
                viewModel,
                currentUser
            )
        }

        viewModel.errorMessage.observe(viewLifecycleOwner) { msg ->
            if (!msg.isNullOrEmpty()) {
                Toast.makeText(context, msg, Toast.LENGTH_SHORT).show()
            }
        }

        viewModel.isVerified.observe(viewLifecycleOwner) { isVerified ->
            viewBinder.ivVerifiedBadge?.let {
                UserVerificationManager.setupVerificationBadge(it, isVerified)
            }
        }

        viewModel.userPosts.observe(viewLifecycleOwner) { posts ->
            val userIsAdmin = AuthManager.isCurrentUserAdmin(requireContext())

            val profilePostListener = object : PostAdapter.PostInteractionListener {
                override fun onUserClick(userId: String) {
                    // Cliccare se stessi sul proprio profilo non fa nulla
                }

                override fun onEditClick(post: Post, position: Int) {
                    viewBinder.rvMyPosts.adapter?.let { adapter ->
                        PostDialogHelper.showEditPostDialog(requireContext(), post, adapter, position)
                    }
                }

                override fun onCommentClick(post: Post, position: Int) {
                    viewBinder.rvMyPosts.adapter?.let { adapter ->
                        PostDialogHelper.showCommentsSheet(requireContext(), post, currentUser, adapter, position)
                    }
                }

                override fun onReportClick(post: Post) {
                    PostDialogHelper.showReportDialog(requireContext(), post, currentUser)
                }

                override fun onShareClick(post: Post, shareContent: String) {
                    val sendIntent = Intent(Intent.ACTION_SEND).apply {
                        putExtra(Intent.EXTRA_TEXT, shareContent)
                        type = "text/plain"
                    }
                    startActivity(Intent.createChooser(sendIntent, "Condividi post tramite:"))
                }

                override fun onImageZoomClick(imageView: View, mediaUri: String) {
                    Zoomy.Builder(requireActivity())
                        .target(imageView)
                        .enableImmersiveMode(false)
                        .animateZooming(true)
                        .register()
                }
            }

            viewBinder.rvMyPosts.adapter = PostAdapter(posts, currentUser, userIsAdmin, profilePostListener)
        }
    }
}